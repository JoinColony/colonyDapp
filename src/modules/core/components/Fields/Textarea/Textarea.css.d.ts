export const container: string;
export const containerHorizontal: string;
export const textareaWrapper: string;
export const TextareaAutoresizeWrapper: string;
export const textareaAutoresizeWrapper: string;
export const main: string;
export const resizableHorizontal: string;
export const resizableVertical: string;
export const resizableBoth: string;
export const layoutInline: string;
export const themeFat: string;
export const alignRight: string;
export const colorSchemaGrey: string;
export const colorSchemaDark: string;
export const colorSchemaTransparent: string;
export const error: string;
export const count: string;
export const limit: string;
