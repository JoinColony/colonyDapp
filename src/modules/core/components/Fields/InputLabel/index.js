/* @flow */

export { default } from './InputLabel';
