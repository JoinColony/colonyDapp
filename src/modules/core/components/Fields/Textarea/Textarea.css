@import 'variables.css';

.main {
  composes: main from '../Field/Field.css';
}

.textarea {
  composes: input from '../Field/Field.css';
  padding: 5px;
}
