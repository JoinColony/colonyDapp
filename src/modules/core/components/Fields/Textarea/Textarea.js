/* @flow */

import TextareaComponent from './Textarea.jsx';
import { asField } from '../Field';

export default asField()(TextareaComponent);
