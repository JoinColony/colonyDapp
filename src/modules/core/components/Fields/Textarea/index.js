/* @flow */

export { default } from './Textarea';
