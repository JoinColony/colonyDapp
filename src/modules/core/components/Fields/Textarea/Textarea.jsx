/* @flow */

import React from 'react';

import styles from './Textarea.css';

import type { FieldComponentProps } from '../flowTypes';

import InputLabel from '../InputLabel';

type CustomProps = {
  disabled?: boolean,
};

type Props = FieldComponentProps<CustomProps>;

const Textarea = ({ elementOnly,
  error,
  hasError,
  help,
  input,
  inputProps,
  passthroughProps: { disabled, ...props },
  label,
}: Props) => (
  <div
    className={styles.main}
    aria-invalid={hasError}
    aria-disabled={disabled}
  >
    {!elementOnly && label ?
      <InputLabel id={inputProps.id} label={label} error={hasError && error} help={help} />
      :
      null
    }
    <textarea
      {...props}
      {...inputProps}
      {...input}
      className={styles.textarea}
    />
  </div>
);

export default Textarea;
