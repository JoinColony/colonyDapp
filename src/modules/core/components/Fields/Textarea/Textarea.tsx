import React, { ReactNode, RefObject, useState } from 'react';
import { MessageDescriptor, useIntl } from 'react-intl';
import { useField } from 'formik';
import cx from 'classnames';
import { nanoid } from 'nanoid';

import { SimpleMessageValues } from '~types/index';
import { getMainClasses } from '~utils/css';

import styles from './Textarea.css';

import InputLabel from '../InputLabel';
import InputStatus from '../InputStatus';

export interface Appearance {
  theme?: 'fat';
  align?: 'right';
  layout?: 'inline';
  resizable?: 'both' | 'horizontal' | 'vertical';
  direction?: 'horizontal';
  colorSchema?: 'dark' | 'transparent' | 'grey';
  size?: 'small';
}

export interface Props {
  /** Appearance object */
  appearance?: Appearance;
  /** Should textarea be displayed alone, or with label & status? */
  elementOnly?: boolean;
  /** Extra node to render on the top right in the label */
  extra?: ReactNode;
  /** Help text */
  help?: string | MessageDescriptor;
  /** Help text values for intl interpolation */
  helpValues?: SimpleMessageValues;
  /** Textarea html `id` attribute */
  id?: string;
  /** Pass a ref to the `<textarea>` element */
  innerRef?: RefObject<HTMLTextAreaElement>;
  /** Input label text */
  label: string | MessageDescriptor;
  /** Input label values for intl interpolation */
  labelValues?: SimpleMessageValues;
  /** Maximum length (will show counter) */
  maxLength?: number;
  /** Textarea field name (form variable) */
  name: string;
  /** Placeholder text */
  placeholder?: string | MessageDescriptor;
  /** Placeholder text values for intl interpolation */
  placeholderValues?: SimpleMessageValues;
  /** Status text */
  status?: string | MessageDescriptor;
  /** Status text values for intl interpolation */
  statusValues?: SimpleMessageValues;
  /** Disabled status of Textarea */
  disabled?: boolean;
  dataTest?: string;
}

const displayName = 'Textarea';

const Textarea = ({
  appearance = {},
  elementOnly = false,
  extra,
  help,
  helpValues,
  id: idProp,
  innerRef,
  label,
  labelValues,
  maxLength = undefined,
  name,
  placeholder: placeholderProp,
  placeholderValues,
  status,
  statusValues,
  disabled,
  dataTest,
}: Props) => {
  const { formatMessage } = useIntl();
  const [id] = useState(idProp || nanoid());
  const [{ value, ...fieldInputProps }, { error }] = useField<string>({
    name,
    value: '',
  });

  const length = value ? value.length : 0;
  const placeholder =
    typeof placeholderProp === 'object'
      ? formatMessage(placeholderProp, placeholderValues)
      : placeholderProp;

  return (
    <div className={styles.container}>
      <InputLabel
        appearance={appearance}
        extra={extra}
        help={help}
        helpValues={helpValues}
        inputId={id}
        label={label}
        labelValues={labelValues}
        screenReaderOnly={elementOnly}
      />
      <div className={styles.textareaWrapper}>
        <textarea
          {...fieldInputProps}
          aria-invalid={error ? true : undefined}
          className={getMainClasses(appearance, styles)}
          id={id}
          maxLength={maxLength}
          placeholder={placeholder}
          ref={innerRef}
          value={value}
          aria-disabled={disabled}
          disabled={disabled}
          data-test={dataTest}
        />
        {maxLength && (
          <span
            className={cx(styles.count, {
              [styles.limit]: length === maxLength,
            })}
          >
            {length}/{maxLength}
          </span>
        )}
      </div>
      {!elementOnly && (
        <InputStatus
          appearance={appearance}
          status={status}
          statusValues={statusValues}
          error={error}
        />
      )}
    </div>
  );
};

Textarea.displayName = displayName;

export default Textarea;
