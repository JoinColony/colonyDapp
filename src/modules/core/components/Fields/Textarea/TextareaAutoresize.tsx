import React, { ReactNode } from 'react';
import Textarea, { TextareaAutosizeProps } from 'react-textarea-autosize';
import { useField } from 'formik';
import { MessageDescriptor, useIntl } from 'react-intl';

import { SimpleMessageValues } from '~types/index';
import { getMainClasses } from '~utils/css';

import styles from './Textarea.css';

import InputLabel from '../InputLabel';
import InputStatus from '../InputStatus';

type Appearance = {
  theme?: 'fat';
  align?: 'right';
  layout?: 'inline';
  resizable?: 'both' | 'horizontal' | 'vertical';
  direction?: 'horizontal';
  colorSchema?: 'dark' | 'grey' | 'transparent';
  size?: 'small';
};

interface Props extends Omit<TextareaAutosizeProps, 'placeholder'> {
  /** Appearance object */
  appearance?: Appearance;

  /** Should the element render with its label or not */
  elementOnly?: boolean;

  /** Extra node to render on the top right in the label */
  extra?: ReactNode;

  /** Help text */
  help?: string | MessageDescriptor;

  /** Help text values for intl interpolation  */
  helpValues?: SimpleMessageValues;

  /** Pass a ref to the `<textarea>` element */
  innerRef?: (ref: HTMLElement | null) => void;

  /** Label text */
  label: string | MessageDescriptor;

  /** Label text values for intl interpolation  */
  labelValues?: SimpleMessageValues;

  /** Html input `name` attribute */
  name: string;

  /** Placeholder text */
  placeholder?: string | MessageDescriptor;

  /** Status text */
  status?: string | MessageDescriptor;

  /** Status text values */
  statusValues?: SimpleMessageValues;

  /** Provides value for data-cy used on cypress testing */
  dataCy?: string;
}

const displayName = 'TextareaAutoresize';

const TextareaAutoresize = ({
  id,
  appearance = {},
  elementOnly,
  help,
  helpValues,
  innerRef,
  extra,
  label,
  labelValues,
  name,
  minRows = 1,
  maxRows = 4,
  placeholder: placeholderProp,
  status,
  statusValues,
  dataCy,
  ...rest
}: Props) => {
  const { formatMessage } = useIntl();
  const [inputFieldProps, { error }] = useField(name);

  const placeholder =
    typeof placeholderProp === 'object'
      ? formatMessage(placeholderProp)
      : placeholderProp;

  const inputProps: TextareaAutosizeProps & {
    ref: ((ref: HTMLElement | null) => void) | undefined;
    'data-cy': string | undefined;
  } = {
    'aria-invalid': error ? 'true' : undefined,
    className: getMainClasses(appearance, styles),
    id,
    maxRows,
    minRows,
    placeholder,
    ref: innerRef,
    'data-cy': dataCy,
    title:
      typeof label === 'object' ? formatMessage(label, labelValues) : label,
    ...rest,
    ...inputFieldProps,
  };

  const renderAutoresizingTextarea = () => {
    return (
      <div className={styles.TextareaAutoresizeWrapper}>
        <Textarea {...inputProps} />
      </div>
    );
  };

  if (elementOnly) {
    return renderAutoresizingTextarea();
  }

  return (
    <div className={styles.container}>
      <InputLabel
        appearance={appearance}
        inputId={id}
        label={label}
        labelValues={labelValues}
        help={help}
        helpValues={helpValues}
        extra={extra}
        screenReaderOnly={elementOnly}
      />
      <div className={styles.TextareaAutoresizeWrapper}>
        {renderAutoresizingTextarea()}
      </div>
      <InputStatus
        appearance={appearance}
        status={status}
        statusValues={statusValues}
        error={error}
      />
    </div>
  );
};

TextareaAutoresize.displayName = displayName;

export default TextareaAutoresize;
