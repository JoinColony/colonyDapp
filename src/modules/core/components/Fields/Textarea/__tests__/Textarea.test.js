/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl } from '$utils/test';

import Textarea from '../Textarea.jsx';
import InputLabel from '../../InputLabel';

jest.mock('../../InputLabel');

describe('Textarea component', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl((
      <Textarea
        meta={{ active: false }}
        elementOnly={false}
        label="foo"
        input={{ id: 'foo' }}
        inputProps={{}}
        passthroughProps={{}}
        help="halp"
        error="red"
        hasError={false}
      />
    ));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Does not wrap with InputLabel without label prop', () => {
    const wrapper = shallowWithIntl((
      <Textarea
        meta={{ active: false }}
        elementOnly={false}
        input={{}}
        inputProps={{ id: 'foo' }}
        passthroughProps={{}}
      />
    ));
    expect(wrapper.find(InputLabel)).not.toBePresent();
  });

  test('Does not wrap with InputLabel if elementOnly is true', () => {
    const wrapper = shallowWithIntl((
      <Textarea
        meta={{ active: false }}
        elementOnly
        label="foo"
        input={{}}
        inputProps={{ id: 'foo' }}
        passthroughProps={{}}
      />
    ));
    expect(wrapper.find(InputLabel)).not.toBePresent();
  });
});
