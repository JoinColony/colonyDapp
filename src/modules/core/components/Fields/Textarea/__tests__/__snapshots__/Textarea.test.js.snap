// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea component Renders initial component 1`] = `
<div
  aria-invalid={false}
>
  <InjectIntl(core.InputLabel)
    error={false}
    help="halp"
    label="foo"
  />
  <textarea
    id="foo"
  />
</div>
`;
